lahja/asyncio/endpoint.py

```diff
@@ -403,12 +422,18 @@ def run(self, *args, **kwargs):  # type: ignore
         # the Endpoint on the other end blocks until it receives this message
         await remote.notify_subscriptions_updated(self.subscribed_events)
 
+        await self._add_half_connection(remote)
+
     async def _handle_client(self, remote: RemoteEndpoint) -> None:
         try:
             async with run_remote_endpoint(remote):
-                await remote.wait_stopped()
+                try:
+                    await remote.wait_stopped()
+                except asyncio.CancelledError:
+                    return
         finally:
-            self._half_connections.remove(remote)
+            if remote in self._half_connections:
+                self._half_connections.remove(remote)
```

Comment:
It might be a little cleaner to replace this finally block with another `task.add_done_callback(self._half_connections.remote(remote))` inside `_accept_conn`.